Flaky CI tests

### Version

latest

### What happened?

For my last commit:
- Ubuntu job was ok, Windows failed.
- On the second run, Ubuntu job failed and Windows was ok.
- On the third run both were ok.

https://github.com/waf/CSharpRepl/runs/4680137856?check_suite_focus=true

In both cases it failed with following:

Run dotnet test --no-build --verbosity normal /p:CollectCoverage=true /p:CoverletOutputFormat=opencover
Build started 1/1/2022 4:38:14 PM.
     1>Project "D:\a\CSharpRepl\CSharpRepl\CSharpRepl.sln" on node 1 (VSTest target(s)).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Debug|Any CPU".
Test run for D:\a\CSharpRepl\CSharpRepl\CSharpRepl.Tests\bin\Debug\net6.0\CSharpRepl.Tests.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Writing trace log to C:\Users\runneradmin\AppData\Local\Temp\tmp569E.tmp


Writing trace log to C:\Users\runneradmin\AppData\Local\Temp\tmp56BE.tmp

[xUnit.net 00:00:49.35]     CSharpRepl.Tests.ReadEvalPrintLoopTests.RunAsync_LoadScript_RunsScript [FAIL]
  Failed CSharpRepl.Tests.ReadEvalPrintLoopTests.RunAsync_LoadScript_RunsScript [14 ms]
  Error Message:
   System.InvalidOperationException : Failed to apply edit to workspace
  Stack Trace:
     at CSharpRepl.Services.Extensions.RoslynExtensions.ApplyChanges(Solution edit, Workspace workspace) in D:\a\CSharpRepl\CSharpRepl\CSharpRepl.Services\Extensions\RoslynExtensions.cs:line 16
   at CSharpRepl.Services.Roslyn.WorkspaceManager.UpdateCurrentDocument(Success result) in D:\a\CSharpRepl\CSharpRepl\CSharpRepl.Services\Roslyn\WorkspaceManager.cs:line 73
   at CSharpRepl.Services.Roslyn.RoslynServices.EvaluateAsync(String input, String[] args, CancellationToken cancellationToken) in D:\a\CSharpRepl\CSharpRepl\CSharpRepl.Services\Roslyn\RoslynServices.cs:line 95
   at CSharpRepl.ReadEvalPrintLoop.Preload(RoslynServices roslyn, IConsole console, Configuration config) in D:\a\CSharpRepl\CSharpRepl\CSharpRepl\ReadEvalPrintLoop.cs:line 102
   at CSharpRepl.ReadEvalPrintLoop.RunAsync(Configuration config) in D:\a\CSharpRepl\CSharpRepl\CSharpRepl\ReadEvalPrintLoop.cs:line 38
   at CSharpRepl.Tests.ReadEvalPrintLoopTests.RunAsync_LoadScript_RunsScript() in D:\a\CSharpRepl\CSharpRepl\CSharpRepl.Tests\ReadEvalPrintLoopTests.cs:line 91
--- End of stack trace from previous location ---

Failed!  - Failed:     1, Passed:    90, Skipped:     0, Total:    91, Duration: 59 s - CSharpRepl.Tests.dll (net6.0)
     1>Done Building Project "D:\a\CSharpRepl\CSharpRepl\CSharpRepl.sln" (VSTest target(s)) -- FAILED.

Build FAILED.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:01:06.13
Error: Process completed with exit code 1.

--- a/CSharpRepl.Services/Roslyn/RoslynServices.cs
+++ b/CSharpRepl.Services/Roslyn/RoslynServices.cs
@@ -32,6 +32,7 @@ public sealed class RoslynServices
 {
     private readonly SyntaxHighlighter highlighter;
     private readonly ITraceLogger logger;
+    private readonly SemaphoreSlim semaphore = new(1);
     private ScriptRunner? scriptRunner;
     private WorkspaceManager? workspaceManager;
     private Disassembler? disassembler;
@@ -84,18 +85,28 @@ public sealed class RoslynServices
     {
         await Initialization.ConfigureAwait(false);
 
-        var result = await scriptRunner
-            .RunCompilation(input.Trim(), args, cancellationToken)
-            .ConfigureAwait(false);
-
-        if (result is EvaluationResult.Success success)
+        try
         {
-            // update our final document text, and add a new, empty project that can be
-            // used for future evaluations (whether evaluation, syntax highlighting, or completion)
-            workspaceManager!.UpdateCurrentDocument(success);
-        }
+            //each RunCompilation (modifies script state) and UpdateCurrentDocument (changes CurrentDocument) cannot be run concurrently
+            await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
 
-        return result;
+            var result = await scriptRunner
+                .RunCompilation(input.Trim(), args, cancellationToken)
+                .ConfigureAwait(false);
+
+            if (result is EvaluationResult.Success success)
+            {
+                // update our final document text, and add a new, empty project that can be
+                // used for future evaluations (whether evaluation, syntax highlighting, or completion)
+                workspaceManager!.UpdateCurrentDocument(success);
+            }
+
+            return result;
+        }
+        finally
+        {
+            semaphore.Release();
+        }
     }
 
     public async Task<string?> PrettyPrintAsync(object? obj, bool displayDetails)

